### DISCUSSION TITLE
Demo to show preferred workflow (mpy main +workers)

### DISCUSSION THREAD

**Discussion by Neon22** [pyscript/pyscript#2193] [Q&A]
In the examples there is no demonstration of what appears to be the suggested primary workflow/organisation of files/code.
We see this discussed over time in several places but no extent demo exists.
Ideally there would be three or four templates which a user could use to start from.

## Ideal workflow seems to be:
1. main.py is an mpy
2. Work that needs more capability is done on workers (pyodide where required else mpy)
3. The workers either directly adj the main's dom or send data back to the main for it to adjust its own dom.

Within this structure its not necessarily clear whether there is a method for moving large amounts of data between js/python so it can be maximally efficient. (e.g. canvas draw calls)
Its also unclear if starting a pyodide worker from inside the `main.mpy` is "better" than starting it in the `index.html`.

So I propose the following templates and an example app to exercise them:

## Templates:
1. main.py(mpy), worker.py(pyo) that directly updates dom id fields.
2. Same but data is sent back to main, with no dom manipulation by the worker.
   - I.e. where main further manipulates and updates the dom itself.
4. More than one Worker (perhaps one fetches some data, and another processes it using a module (like the many pandas examples))

## Example problem
To make an interesting but useful demo I suggest the following structure/tasks:

1. main.py(mpy) has buttons and a canvas
2. main asks a worker to build a datastructure (the two-worker variant has it getting this data from a fetch worker)
3. main asks the worker to manipulate this data and create `svg` elements.
4. main needs these in an `Image` format to draw them on the canvas
4a. the worker makes these into images and draws them directly on the canvas, or
4b. the images get transferred back to main which draws them on the canvas
5. main ui buttons enable initiating the two worker actions and doing other canvas ops locally.
5a. Could add timing if we wanted to see if one path or another was faster.

A demo like this would show:
- several examples of increasing complexity (2 workers, js data transfer, dom manip in worker, fetch example, worker sync.)
- something with responsive visual feedback
- when we need to show wait animations and how to handshake waiting until the worker is 'ready to run' / 'data is ready'

There is a start at this here:
 - https://pyscript.com/@neon22/mpy-master-template/latest?files=main.py

**Comment by Neon22** [comment]
OK. first problem is that micropython does not appear to work with the recipes for file upload and download.
The `io.BytesIO` doesn't give expected results and neither does `js.Uint8Array`

This means that having mpy as the main.py means using pyodide workers for a lot more tasks. I assume each will start slow. So maybe this idea of using mpy is not as good as just having a slow starting pyodide main and less workers.

Any thoughts on this ?

**Reply by WebReflection** [reply]
we convert with relative ease JS views into M/PY memoriviews but we are not exposing these utilities yet ... once we do, I believe it should be relatively trivial to upload files also because I've done it already "*here and there*" with live demoes and whatnot (mpy on main driving MicroPython microcontrollers, as example, where I can upload, read or write buffers in there too). I will try to come up with some example that satisfies your requirements but it'd be great to have a checkbox based list so I can read, implement, and flag all of these (or tackle these via future MRs).

**Reply by Neon22** [reply]
I'll make a list and put it here.

by **recipes** I was referring to the JeffGlass upload and download recipes here:
 - https://pyscript.recipes/2024.5.2/basic/file-download/
 - they are in common use
 - However I see  them as stopgap measures because the Filesystem API seems to be the standard to implement and it enables a lot more usecases than the simple recipe.
    - https://developer.mozilla.org/en-US/docs/Web/API/File_System_API

Chris Laffra's demo of the slowdown inherent in JS vs python argument conversion (below)
 - is a good example of where maybe some support could be offered in the to_js() kind of approach but I am not fluent here.

Problems I need to solve:
 - to get an svg element (created not loaded) drawn on a canvas requires using an Image() as an intermediary and setting its src attribute then loading it via createObjectURL()  or alternatively as a Blob().
 - to efficiently draw a number of svg fragments means making a number of images, perhaps transferring them to the JS side and then drawing them onto the canvas. Being able to do it at all would be great, Doing it efficiently would be fantastic.

**Comment by Neon22** [comment]
## Javascript/python overhead:

Chris Laffra has a demo showing python_javascript argument conversion overhead here:
 - https://pyscript.com/@neon22/ref-buffered-canvas-drawing/latest
 - It shows that jsondumps is a good way to bundle up lots of calls to send from python side to javascript side to actually do drawing.
 - 2.7X faster (if chrome devtools cleared)
